lib/improver/ensemble_calibration.py

```diff
@@ -0,0 +1,1418 @@
+            current_forecast_cubes)
+        historic_forecast_cubes = concatenate_cubes(
+            historic_forecast_cubes)
+        truth_cubes = concatenate_cubes(truth_cubes)
+
+        for current_forecast_cube in current_forecast_cubes.slices_over(
+                "time"):
+            date = unit.num2date(
+                current_forecast_cube.coord("time").points,
+                current_forecast_cube.coord("time").units.name,
+                current_forecast_cube.coord("time").units.calendar)[0]
+            # Extract desired forecast_period from historic_forecast_cubes.
+            forecast_period_constr = iris.Constraint(
+                forecast_period=current_forecast_cube.coord(
+                    "forecast_period").points)
+            historic_forecast_cube = historic_forecast_cubes.extract(
+                forecast_period_constr)
+
+            # Extract truth matching the time of the historic forecast.
+            truth_constr = iris.Constraint(
+                forecast_reference_time=historic_forecast_cube.coord(
+                    "time").points)
+            truth_cube = truth_cubes.extract(truth_constr)
+
+            if truth_cube is None:
+                msg = ("Unable to calibrate for the time points {} "
+                       "as no truth data is available."
+                       "Moving on to try to calibrate "
+                       "next time point.".format(
+                           historic_forecast_cube.coord("time").points))
+                warnings.warn(msg)
+                continue
+
+            # Make sure inputs have the same units.
+            historic_forecast_cube.convert_units(self.desired_units)
+            truth_cube.convert_units(self.desired_units)
+
+            if self.predictor_of_mean_flag.lower() in ["mean"]:
+                no_of_members = None
+                forecast_predictor = historic_forecast_cube.collapsed(
+                    "realization", iris.analysis.MEAN)
+            elif self.predictor_of_mean_flag.lower() in ["members"]:
+                no_of_members = len(
+                    historic_forecast_cube.coord("realization").points)
+                forecast_predictor = historic_forecast_cube
+
+            forecast_var = historic_forecast_cube.collapsed(
+                "realization", iris.analysis.VARIANCE)
+
+            # Computing initial guess for EMOS coefficients
+            # If no initial guess from a previous iteration, or if there
+            # are NaNs in the initial guess, calculate an initial guess.
+            if "initial_guess" not in locals() or nan_in_initial_guess:
+                initial_guess = self.compute_initial_guess(
+                    truth_cube, forecast_predictor,
+                    self.predictor_of_mean_flag,
+                    self.ESTIMATE_COEFFICIENTS_FROM_LINEAR_MODEL_FLAG,
+                    no_of_members=no_of_members)
+
+            if np.any(np.isnan(initial_guess)):
```

Comment:
For the first iteration, `initial_guess` is not defined, so the value for the initial_guess is only checked for NaN values after it has been calculated for the first time.